NPE: ColumnIndices.getColumnInfo(java.lang.Class) after upgrading to 0.84.0

Hi team,

After upgrading from 0.82.2 to 0.84.0 got this exception during the first app start:
```
java.lang.NullPointerException: Attempt to invoke virtual method 'io.realm.internal.ColumnInfo io.realm.internal.ColumnIndices.getColumnInfo(java.lang.Class)' on a null object reference
```

Didn't happen during the next starts.
Same happened when I deployed the app on another device with the same 0.82.2 realm version.

Here is the piece of code which triggered the crash on the first device (called on the worker thread):
```java
realm.where(PinRealm.class)
                    .equalTo("token", pinToken)
                    .findFirst();
```

And here is the piece of code which triggered the crash on another device (called on the main thread):
```java
realm.where(PinRealm.class)
                    .equalTo("deleted", false)
                    .equalTo("inboxState", NEW.ordinal())
                    .findAll();
```

Same NPE. 

```
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'io.realm.internal.ColumnInfo io.realm.internal.ColumnIndices.getColumnInfo(java.lang.Class)' on a null object reference
       at io.realm.Realm.getColumnInfo(Realm.java:727)
       at io.realm.RealmMyClassRealmProxy.copyOrUpdate(RealmMyClassRealmProxy.java:1671)
       at io.realm.RealmModuleMyClassMediator.copyOrUpdate(RealmModuleMyClassMediator.java:97)
       at io.realm.Realm.copyOrUpdate(Realm.java:1130)
       at io.realm.Realm.copyToRealmOrUpdate(Realm.java:763)
       at com.fictionpress.fanfiction.realm.table.RealmTableMyClass$1$1.execute(RealmTableMyClass.java:110)
       at io.realm.Realm.executeTransaction(Realm.java:1034)
       at com.fictionpress.fanfiction.realm.table.RealmTableMyClass$1.run(RealmTableMyClass.java:107)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
```

Using ---> to note crash line.

```java
    public static RealmMyClass copyOrUpdate(Realm realm, RealmMyClass object, boolean update, Map<RealmObject,RealmObjectProxy> cache) {
        if (object.realm != null && object.realm.getPath().equals(realm.getPath())) {
            return object;
        }
        RealmMyClass realmObject = null;
        boolean canUpdate = update;
        if (canUpdate) {
            Table table = realm.getTable(RealmMyClass.class);
            long pkColumnIndex = table.getPrimaryKey();
            long rowIndex = table.findFirstLong(pkColumnIndex, object.getStoryId());
            if (rowIndex != TableOrView.NO_MATCH) {
                ------> realmObject = new RealmMyClassRealmProxy(realm.getColumnInfo(RealmMyClass.class));
                realmObject.realm = realm;
                realmObject.row = table.getUncheckedRow(rowIndex);
                cache.put(object, (RealmObjectProxy) realmObject);
            } else {
                canUpdate = false;
            }
        }
```

--- a/realm/src/androidTest/java/io/realm/RealmTest.java
+++ b/realm/src/androidTest/java/io/realm/RealmTest.java
@@ -16,6 +16,7 @@
 package io.realm;
 
 import android.content.Context;
+import android.os.SystemClock;
 import android.test.AndroidTestCase;
 
 import junit.framework.AssertionFailedError;
@@ -44,7 +45,6 @@ import java.util.concurrent.TimeoutException;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import io.realm.dynamic.DynamicRealmObject;
-import io.realm.entities.AllJavaTypes;
 import io.realm.entities.AllTypes;
 import io.realm.entities.AllTypesPrimaryKey;
 import io.realm.entities.AnnotationIndexTypes;
@@ -2218,5 +2218,52 @@ public class RealmTest extends AndroidTestCase {
         assertTrue(testRealm.isInTransaction());
         testRealm.cancelTransaction();
         assertFalse(testRealm.isInTransaction());
+    }
+
+    // test for https://github.com/realm/realm-java/issues/1646
+    public void testClosingRealmWhileOtherThreadIsOpeningRealm() throws Exception {
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch endLatch = new CountDownLatch(1);
+
+        final List<Exception> exception = new ArrayList<Exception>();
+
+        new Thread() {
+            @Override
+            public void run() {
+                try {
+                    startLatch.await();
+                } catch (InterruptedException e) {
+                    exception.add(e);
+                    return;
+                }
+
+                final Realm realm = Realm.getInstance(testConfig);
+                try {
+                    realm.where(AllTypes.class).equalTo("columnLong", 0L).findFirst();
+                } catch (Exception e) {
+                    exception.add(e);
+                } finally {
+                    endLatch.countDown();
+                    realm.close();
+                }
+            }
+        }.start();
+
+        // prevent for another thread to enter Realm.createAndValidate().
+        synchronized (BaseRealm.class) {
+            startLatch.countDown();
+
+            // wait for another thread's entering Realm.createAndValidate().
+            SystemClock.sleep(100L);
+
+            testRealm.close();
+            testRealm = null;
+        }
+
+        endLatch.await();
+
+        if (!exception.isEmpty()) {
+            throw exception.get(0);
+        }
     }
 }

--- a/realm/src/main/java/io/realm/Realm.java
+++ b/realm/src/main/java/io/realm/Realm.java
@@ -243,8 +243,7 @@ public final class Realm extends BaseRealm {
     private static synchronized Realm create(RealmConfiguration configuration) {
         boolean autoRefresh = Looper.myLooper() != null;
         try {
-            boolean validateSchema = !validatedRealmFiles.containsKey(configuration.getPath());
-            return createAndValidate(configuration, validateSchema, autoRefresh);
+            return createAndValidate(configuration, null, autoRefresh);
 
         } catch (RealmMigrationNeededException e) {
             if (configuration.shouldDeleteRealmIfMigrationNeeded()) {
@@ -257,8 +256,12 @@ public final class Realm extends BaseRealm {
         }
     }
 
-    private static Realm createAndValidate(RealmConfiguration configuration, boolean validateSchema, boolean autoRefresh) {
+    private static Realm createAndValidate(RealmConfiguration configuration, Boolean validateSchema, boolean autoRefresh) {
         synchronized (BaseRealm.class) {
+            if (validateSchema == null) {
+                validateSchema = !validatedRealmFiles.containsKey(configuration.getPath());
+            }
+
             // Check if a cached instance already exists for this thread
             String canonicalPath = configuration.getPath();
             Map<RealmConfiguration, Integer> localRefCount = referenceCount.get();
@@ -1149,7 +1152,10 @@ public final class Realm extends BaseRealm {
 
     @Override
     protected void lastLocalInstanceClosed() {
-        validatedRealmFiles.remove(configuration.getPath());
+        // validatedRealmFiles must not modified while other thread is executing createAndValidate()
+        synchronized (BaseRealm.class) {
+            validatedRealmFiles.remove(configuration.getPath());
+        }
         realmsCache.get().remove(configuration);
     }
 

